util/zenfs.cc

```diff
@@ -69,22 +70,78 @@ Status zenfs_mount(std::unique_ptr<ZonedBlockDevice> &zbd,
   return s;
 }
 
-int zenfs_tool_mkfs() {
-  Status s;
+int is_dir(const char *path) {
+  struct stat st;
+  if (stat(path, &st) != 0) {
+    fprintf(stderr, "Failed to stat %s\n", path);
+    return 1;
+  }
+  return S_ISDIR(st.st_mode);
+}
+
+// Create or check pre-existing aux directory and fail if it is
+// inaccessible by current user and if it has previous data
+int create_aux_dir(const char *path) {
+  struct dirent *dent;
   DIR *aux_dir;
+  size_t nfiles = 0;
+  int ret = 0;
+
+  errno = 0;
+  ret = mkdir(FLAGS_aux_path.c_str(), 0750);
+  if (ret < 0 && EEXIST != errno) {
+    fprintf(stderr, "Failed to create aux directory %s: %s\n",
+            FLAGS_aux_path.c_str(), strerror(errno));
+    return 1;
+  }
+  // The aux_path is now available, check if it is a directory infact
+  // and is empty and the user has access permission
 
-  if (FLAGS_aux_path.empty()) {
-    fprintf(stderr, "You need to specify --aux_path\n");
+  if (!is_dir(path)) {
+    fprintf(stderr, "Aux path %s is not a directory\n", path);
     return 1;
   }
 
+  errno = 0;
   aux_dir = opendir(FLAGS_aux_path.c_str());
-  if (ENOENT != errno) {
-    fprintf(stderr, "Error: aux path exists\n");
+  if (errno) {
+    fprintf(stderr, "Failed to open aux directory %s: %s\n",
+            FLAGS_aux_path.c_str(), strerror(errno));
+    closedir(aux_dir);
+    return 1;
+  }
+
+  // Consider the directory as non-empty if any files/dir other
+  // than . and .. are found.
+  while ((dent = readdir(aux_dir)) != NULL && nfiles <= 2) ++nfiles;
+  if (nfiles > 2) {
+    fprintf(stderr, "Aux directory %s is not empty.\n", FLAGS_aux_path.c_str());
     closedir(aux_dir);
     return 1;
   }
 
+  if (access(FLAGS_aux_path.c_str(), R_OK | W_OK | X_OK) < 0) {
+    fprintf(stderr,
+            "User does not have access permissions on "
+            "aux directory %s\n",
+            FLAGS_aux_path.c_str());
+    closedir(aux_dir);
+    return 1;
+  }
+
+  return 0;
```

Comment:
I think we need `closedir()` here as well, right?

Reply:
You are right, thanks, added it.

Reply:
@aravind-wdc This is exactly why I suggested using RAII, instead of 3 manual calls to `closedir()` before each `return` (one of which was missing), you declare everything in one place and delegate the responsibility of releasing allocated resources to the compiler.

Reply:
@percona-ysorokin completely agree, this is a perfect demonstration of why RAII is awesome :)